[benchmark] Variants for Set Sequence Methods

Gumichocopengin8 opened the pull request:
Added benchmarks for Set.
To support this: https://github.com/apple/swift/pull/24156#pullrequestreview-229848684 
> Some Set operations have generic variants that take a sequence -- e.g., `Set.isSuperset<S: Sequence>(of: S)`. These are currently not covered by benchmarks at all, even though we're planning to [improve them](https://github.com/apple/swift/pull/21300) soon! I'd very much welcome benchmarks that measure those.

Gumichocopengin8 commented:
CC: @lorentey 
Thank you.

Gumichocopengin8 commented:
Do we also need to implement `formUnion`, `formIntersection` and stuff like that?

lorentey commented on benchmark/single-source/SetTests.swift:
```diff
@@ -139,6 +139,88 @@ public let SetTests = [
+    setUpFunction: { blackHole([setOAB, setOCD]) }),
+  BenchmarkInfo(
+    name: "Set.isStrictSuperset.Int25",
+    runFunction: { n in run_SetIsStrictSupersetInt(setB, setAB, false, 50 * n) },
```
It would make sense to swap the arguments here -- we want the `isStrictSuperset` test to succeed so that we actually exercise some containment checks.

lorentey commented on benchmark/single-source/SetTests.swift:
```diff
@@ -139,6 +139,88 @@ public let SetTests = [
+    setUpFunction: { blackHole([setB, setAB]) }),
+  BenchmarkInfo(
+    name: "Set.isStrictSuperset.Box25",
+    runFunction: { n in run_SetIsStrictSupersetBox(setOB, setOAB, false, 50 * n) },
```
Same here -- let's swap `setOB` and `setOAB`, and switch the expected value to true.

lorentey commented on benchmark/single-source/SetTests.swift:
```diff
@@ -139,6 +139,88 @@ public let SetTests = [
+    setUpFunction: { blackHole([setOB, setOAB]) }),
+  BenchmarkInfo(
+    name: "Set.isStrictSuperset.Int50",
+    runFunction: { n in run_SetIsStrictSupersetInt(setY, setXY, false, 50 * n) },
```
And here!

lorentey commented on benchmark/single-source/SetTests.swift:
```diff
@@ -398,6 +597,19 @@ public func run_SetSymmetricDifferenceInt(
+  _ n: Int) {
+  for _ in 0 ..< n {
+    var diff = a
+    diff.formSymmetricDifference(identity(b))
```
This and the other `run_SetForm...` tests aren't quite right: in each iteration they all force a COW copy before the actual operation, and that copy can take longer to execute than the operation itself. So I expect these tests are mostly benchmarks of `Set`'s copying performance, not set operations!

Unfortunately I don't think there is an easy way to properly benchmark these in the current benchmark setup -- we can't move overhead that gets repeated on every single iteration out of the timing loop.

lorentey commented on benchmark/single-source/SetTests.swift:
```diff
@@ -139,6 +139,88 @@ public let SetTests = [
+    setUpFunction: { blackHole([setP, setQ]) }),
+
+  BenchmarkInfo(
+    name: "Set.isStrictSuperset.Empty.Int",
```
Similarly, `isStrictSuperset` is basically the same as `isStrictSubset`. I'm not convinced it needs to be benchmarked separately.

lorentey reviewed (commented):
Hm, I'm not sure we need these benchmarks.

* The non-generic `isSuperset`/`isStrictSuperset` are trivial wrappers around `isSubset`/`isStrictSubset`, and I don't expect they will ever change. (Or rather, they will remain mostly in sync with `isSubset` when they do.)

* The mutation tests don't seem too useful in their current form, since they include a COW copy on every iteration, and I expect that will often swamp the actual operation.

Some Set operations have generic variants that take a sequence -- e.g., `Set.isSuperset<S: Sequence>(of: S)`. These are currently not covered by benchmarks at all, even though we're planning to [improve them](https://github.com/apple/swift/pull/21300) soon! I'd very much welcome benchmarks that measure those.

lorentey commented:
@swift-ci benchmark

Gumichocopengin8 commented:
Thank you for reviewing.
I also wondered if `isSuperset/isStrictSuperset` and mutation functions are not necessary because they are almost the same as `isSubset/isStrictSubset`. Thanks for the confirmation. I'll revert them.

Then, I'll implement benchmark of `Set.isSuperset<S: Sequence>(of: S)` and more.

Gumichocopengin8 commented:
@lorentey 
To implement `Set.isSuperset<S: Sequence>(of: S)`, should I use array or sequence? 
I'm going to use array instead of sequence because sequence doesn't  support `reserveCapacity`.
If there's problem to use array, please give me your advice.

Gumichocopengin8 commented:
@lorentey 
Could you please review this?

Gumichocopengin8 commented on benchmark/single-source/SetTests.swift:
```diff
@@ -55,6 +55,36 @@ let setQ: Set<Int> = {
   return set
 }()
 
+// Construction of empty array.
```
I made arrays as sequences.
Are they ok?

palimondo replied:
I have initially thought you could have gone directly with ranges, which conform to `Sequence`, but for all the `Box` variants it is better to materialize the objects into arrays, otherwise the benchmark would have been dominated by reference counting traffic… so I think using arrays like you did is OK.

Gumichocopengin8 commented on benchmark/single-source/SetTests.swift:
```diff
@@ -55,6 +55,36 @@ let setQ: Set<Int> = {
+let arrayP = Array(0 ..< size)
+let arrayQ: Array<Int> = {
+  var array = Array(0 ..< size)
+  array.shuffle()
```
When I used `reserveCapacity`, the array wasn't random, so I use `shuffle` instead.
Is it the best way for this?

Gumichocopengin8 commented on benchmark/single-source/SetTests.swift:
```diff
@@ -334,6 +610,98 @@ public let SetTests = [
+
+  BenchmarkInfo(
+    name: "Set.Filter.Int0",
+    runFunction: { n in run_SetFilterInt(setAB, size, 0, 10 * n)},
```
Should I change `10 * n` instead of `1000 * n`?

palimondo replied:
I don't understand what are these `Filter` benchmark variants doing here at all… @lorentey requested adding only benchmarks for generic variants of set methods that take `Sequence`, as argument, as far as I've seen.

palimondo replied:
@lorentey Did you request filter benchmark for `Set`, too? If so, @Gumichocopengin8, you might want to look at [DictionaryCopy.swift](https://github.com/apple/swift/blob/master/benchmark/single-source/DictionaryCopy.swift#L35-L42), which stresses `Dictionary` construction with a particular emphasis on avoiding possible a quadratic performance degradation.

Gumichocopengin8 replied:
> Some Set operations have generic variants that take a sequence -- e.g., `Set.isSuperset<S: Sequence>(of: S)`. These are currently not covered by benchmarks at all, even though we're planning to [improve them](https://github.com/apple/swift/pull/21300) soon! I'd very much welcome benchmarks that measure those.

My understanding is that he also request filter benchmark because filter is also in the list.

Gumichocopengin8 commented:
@lorentey @palimondo @airspeedswift Could you please review this?

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -252,6 +497,37 @@ public let SetTests = [
     setUpFunction: { blackHole([setP, setQ]) }),
 
+  BenchmarkInfo(
+    name: "Set.Intersection.Sequence.Int0",
```
Lowercase `i`, intersection is a method name.
```suggestion
    name: "Set.intersection.Seq.Int0",
```
… also in all `intersection` benchmarks below.

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -334,6 +610,98 @@ public let SetTests = [
+    tags: [.validation, .api, .Set],
+    setUpFunction: { blackHole([setOAB, arrayOE]) }),
+  BenchmarkInfo(
+    name: "Set.Subtracting.Sequence.Int0",
```
Lowercase `s`, same as above, `subtracting` is method name.
```suggestion
    name: "Set.subtracting.Seq.Int0",
```
...and same for similar cases below 👇

palimondo reviewed (changes requested):
Could you please use a shortened  `Seq` to designate the sequence variants? I.e. `Set.isStrictSuperset.Seq.Box25`. (Just for consistency with a planned future rename of other `Sequence` benchmarks that will be using `Seq` to refer to the type, as illustrated in the examples under section **Use groups and variants** in the [Naming Convention](https://github.com/apple/swift/blob/master/benchmark/Naming.md)).

I don't see what are the `Filter` benchmark variants doing here at all.

It seems to me there are still missing benchmarks for 2 generic methods that take a `Sequence` parameter:

* `union`
* `symmetricDifference`

I'm assuming @lorentey wants to add these too...

Gumichocopengin8 commented:
@palimondo Thank you for reviewing. I fixed code.
Could you please review it again?

palimondo commented:
Oh, OK, I understand now that this set of new benchmarks should be tailored specifically for the pending improvements in #21300. Then it makes sense not to include `symmetricDifference` and `union` families because they are not affected by those changes, even though a generic `Sequence` consuming variant exists in the `Set` API. Sorry @Gumichocopenguin8, could you please remove them again and edit the PR description to clearly document the purpose of these benchmarks?

As for the `filter` family, I still think it would make sense to take inspiration from [DictionaryCopy](https://github.com/apple/swift/blob/master/benchmark/single-source/DictionaryCopy.swift#L35-L42) and do the four 16k–28k variants for each of `Int50` (filter matches only even numbers) and `Int100` (filter matches all) and leave out the other variants.

So we’d have `Set.filter.Int100.16k` (`.20k`, `.24k`, `.28k`) with “match all” filter and `Set.filter.Int50.16k` (etc.) with “match even” filter.

What do you think @lorentey?

Gumichocopengin8 commented:
@palimondo 
I fixed `filter` benchmark, and delete `symmetricDifference` and `union `. 
I also added `filter.Box100` and `filter.Box50`. 
I believe `box` is significant in this case. 
Could you review this?

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

Gumichocopengin8 commented:
I got warnings.
Should I fix them all?
If so, how can you usually fix them?

palimondo commented:
> I got warnings. Should I fix them all?

**Not yet.** The memory warnings are safe to ignore in these cases. But we need to discuss the surprises…

In principle, it is not a showstopper when some benchmarks are over the recommended runtime threshold  of 1000 μs when there is pending improvement which will bring the runtime down. Going over 10 000 μs (10 ms) is much more serious issue, because this crosses the scheduler quantum on mac os — this means that every sample is guaranteed to include measurement error from a context switch which leads to very noisy results with sporadic false changes in our reports.

We should aim for these benchmarks to run in reasonable times after #21300 lands, buts some of these results are disturbingly high. So high that performance improvements @lorentey mentions in  #21300 will not get the runtimes of these benchmark to the ranges we need for stable measurements. I'm guessing that the sequence iterators are not specialized? This needs more investigation. It might be issue with how the benchmarks are written or there might be much more fundamental issue with resilient API after ABI stability. I'll try to profile these locally.

I don't really see the need to add `filter.Box` variants, as they do not add new information on top of what `filter.Int` does (they are all just a constant slowdown of 0.36x), but that is ultimately @lorentey's call.

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -49,11 +49,7 @@ let setY = Set(half ..< size)         // 200 ..< 400
-  set.reserveCapacity(2 * size)
-  return set
-}()
+let setQ = Set(0 ..< size)
```
The whole point of having sets `P` and `Q` with identical contents but different capacity is described in the above comment.

The `setQ` should not be changed, please revert this part back to as it was. My previous [review comments](https://github.com/apple/swift/pull/24156#discussion_r280842909) were specific to `arrayQ` only.

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -49,12 +49,50 @@ let setY = Set(half ..< size)         // 200 ..< 400
+// Two arrays with 100% overlap, but different bucket counts (let's not make it
+// too easy...)
+let arrayP = Array(0 ..< size)
+let arrayQ: Array<Int> = {
```
As I [said before](https://github.com/apple/swift/pull/24156#discussion_r280842909), we don't need a shuffled `arrayQ` at all. All `Int100` comparisons between a `Set` and a `Sequence` can be performed on `setP` and `arrayP` only. I'll point out all the places below.

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -139,6 +218,129 @@ public let SetTests = [
+    setUpFunction: { blackHole([setXY, arrayY]) }),
+  BenchmarkInfo(
+    name: "Set.isSuperset.Seq.Int100",
+    runFunction: { n in run_SetIsSupersetSeqInt(setQ, arrayP, true, 50 * n) },
```
```suggestion
    runFunction: { n in run_SetIsSupersetSeqInt(setP, arrayP, true, 50 * n) },
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -139,6 +218,129 @@ public let SetTests = [
+    name: "Set.isSuperset.Seq.Int100",
+    runFunction: { n in run_SetIsSupersetSeqInt(setQ, arrayP, true, 50 * n) },
+    tags: [.validation, .api, .Set],
+    setUpFunction: { blackHole([setQ, arrayP]) }),
```
```suggestion
    setUpFunction: { blackHole([setP, arrayP]) }),
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -139,6 +218,129 @@ public let SetTests = [
+    setUpFunction: { blackHole([setXY, arrayY]) }),
+  BenchmarkInfo(
+    name: "Set.isStrictSuperset.Seq.Int100",
+    runFunction: { n in run_SetIsStrictSupersetSeqInt(setQ, arrayP, false, 50 * n) },
```
```suggestion
    runFunction: { n in run_SetIsStrictSupersetSeqInt(setP, arrayP, false, 50 * n) },
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -139,6 +218,129 @@ public let SetTests = [
+    name: "Set.isStrictSuperset.Seq.Int100",
+    runFunction: { n in run_SetIsStrictSupersetSeqInt(setQ, arrayP, false, 50 * n) },
+    tags: [.validation, .api, .Set],
+    setUpFunction: { blackHole([setQ, arrayP]) }),
```
```suggestion
    setUpFunction: { blackHole([setP, arrayP]) }),
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -190,6 +392,57 @@ public let SetTests = [
+    setUpFunction: { blackHole([setY, arrayXY]) }),
+  BenchmarkInfo(
+    name: "Set.isDisjoint.Seq.Int100",
+    runFunction: { n in run_SetIsDisjointSeqInt(setP, arrayQ, false, 5000 * n) },
```
```suggestion
    runFunction: { n in run_SetIsDisjointSeqInt(setP, arrayP, false, 5000 * n) },
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -190,6 +392,57 @@ public let SetTests = [
+    name: "Set.isDisjoint.Seq.Int100",
+    runFunction: { n in run_SetIsDisjointSeqInt(setP, arrayQ, false, 5000 * n) },
+    tags: [.validation, .api, .Set],
+    setUpFunction: { blackHole([setP, arrayQ]) }),
```
```suggestion
    setUpFunction: { blackHole([setP, arrayP]) }),
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -98,6 +136,47 @@ public let SetTests = [
+    setUpFunction: { blackHole([setY, arrayXY]) }),
+  BenchmarkInfo(
+    name: "Set.isSubset.Seq.Int100",
+    runFunction: { n in run_SetIsSubsetSeqInt(setP, arrayQ, true, 50 * n) },
```
```suggestion
    runFunction: { n in run_SetIsSubsetSeqInt(setP, arrayP, true, 50 * n) },
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -98,6 +136,47 @@ public let SetTests = [
+    name: "Set.isSubset.Seq.Int100",
+    runFunction: { n in run_SetIsSubsetSeqInt(setP, arrayQ, true, 50 * n) },
+    tags: [.validation, .api, .Set],
+    setUpFunction: { blackHole([setP, arrayQ]) }),
```
```suggestion
    setUpFunction: { blackHole([setP, arrayP]) }),
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -139,6 +218,129 @@ public let SetTests = [
+    setUpFunction: { blackHole([setY, arrayXY]) }),
+  BenchmarkInfo(
+    name: "Set.isStrictSubset.Seq.Int100",
+    runFunction: { n in run_SetIsStrictSubsetSeqInt(setP, arrayQ, false, 50 * n) },
```
```suggestion
    runFunction: { n in run_SetIsStrictSubsetSeqInt(setP, arrayP, false, 50 * n) },
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -139,6 +218,129 @@ public let SetTests = [
+    name: "Set.isStrictSubset.Seq.Int100",
+    runFunction: { n in run_SetIsStrictSubsetSeqInt(setP, arrayQ, false, 50 * n) },
+    tags: [.validation, .api, .Set],
+    setUpFunction: { blackHole([setP, arrayQ]) }),
```
```suggestion
    setUpFunction: { blackHole([setP, arrayP]) }),
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -252,6 +505,37 @@ public let SetTests = [
+    setUpFunction: { blackHole([setXY, arrayYZ]) }),
+  BenchmarkInfo(
+    name: "Set.intersection.Seq.Int100",
+    runFunction: { n in run_SetIntersectionSeqInt(setP, arrayQ, size, 10 * n) },
```
```suggestion
    runFunction: { n in run_SetIntersectionSeqInt(setP, arrayP, size, 10 * n) },
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -252,6 +505,37 @@ public let SetTests = [
+    name: "Set.intersection.Seq.Int100",
+    runFunction: { n in run_SetIntersectionSeqInt(setP, arrayQ, size, 10 * n) },
+    tags: [.validation, .api, .Set],
+    setUpFunction: { blackHole([setP, arrayQ]) }),
```
```suggestion
    setUpFunction: { blackHole([setP, arrayP]) }),
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -334,6 +618,139 @@ public let SetTests = [
+    setUpFunction: { blackHole([setXY, arrayYZ]) }),
+  BenchmarkInfo(
+    name: "Set.subtracting.Seq.Int100",
+    runFunction: { n in run_SetSubtractingSeqInt(setP, arrayQ, 0, 10 * n) },
```
```suggestion
    runFunction: { n in run_SetSubtractingSeqInt(setP, arrayP, 0, 10 * n) },
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -334,6 +618,139 @@ public let SetTests = [
+    name: "Set.subtracting.Seq.Int100",
+    runFunction: { n in run_SetSubtractingSeqInt(setP, arrayQ, 0, 10 * n) },
+    tags: [.validation, .api, .Set],
+    setUpFunction: { blackHole([setP, arrayQ]) }),
```
```suggestion
    setUpFunction: { blackHole([setP, arrayP]) }),
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -446,6 +959,44 @@ public func run_SetIsDisjointInt(
+@inline(never)
+public func run_SetFilterInt50(_ n: Int) {
+  for _ in 0 ..< n {
+    let copy = set.filter { $0 % 2 == 0 }
```
To me the name `copy` makes sense only in case when the filter matches all elements… How about this?
```suggestion
    let half = set.filter { $0 % 2 == 0 }
```

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -446,6 +959,44 @@ public func run_SetIsDisjointInt(
+public func run_SetFilterInt50(_ n: Int) {
+  for _ in 0 ..< n {
+    let copy = set.filter { $0 % 2 == 0 }
+    CheckResults(set.count == copy.count * 2)
```
```suggestion
    CheckResults(set.count == half.count * 2)
```

palimondo reviewed (changes requested):
I still have a few minor points about `Int100` comparisons and a nitpick about variable naming in filter case.

palimondo reviewed (changes requested):
The reason for those abnormally high runtimes are incorrect multipliers. In this benchmark family, we are using `5000`  only for `Int0`, `Box0` and `Empty` to inflate the workload for the case where the operation is almost nothing — to make it observable. All other cases use `50`. 

This works for set-to-set comparisons in both directions, because their size is known. The size of a `Sequence` is unknown before it is fully consumed. So we need to handle the set-to-sequence `Int0` and `Box0` variants depending on their runtime. 50 is better multiplier constant for those cases.

Edit: I forgot to comment on these, which should also use 50:
* `Set.isSubset.Seq.[Int0, Box0]`
* `Set.isStrictSubset.Seq.[Int0, Box0]`
* `Set.isStrictSuperset.Seq.[Int0, Box0]`

palimondo commented:
These two runtime outliers are interesting:
* `Set.isStrictSubset.Seq.Empty.Int`
* `Set.isStrictSuperset.Seq.Int.Empty`
because they indicate we might benefit from special casing the empty set in the `_NativeSet` overrides. @lorentey?

palimondo commented:
@Gumichocopengin8 Tip: You can apply the [patch from PR 21300](https://github.com/apple/swift/pull/21300.patch) locally to verify the final runtimes will fall, if possible into the 20–1000 μs range (adjusting for your machine's relative performance vis-a-vis the Mac Pros that are running the CI here).

Gumichocopengin8 commented:
@palimondo 
I fixed and adjusted benchmark for PR 21300. 

>Edit: I forgot to comment on these, which should also use 50:
Set.isSubset.Seq.[Int0, Box0]
Set.isStrictSubset.Seq.[Int0, Box0]
Set.isStrictSuperset.Seq.[Int0, Box0]

Basically, I followed your instruction above. However, I didn't apply it sometimes in order to adjust benchmark for PR 21300.

palimondo reviewed (commented):
Basically LGTM.

Thank you for your patience and endurance @Gumichocopengin8, you are going great work here! I have one last naming nitpick about the filter benchmarks below and a little bug fix request. Finally, if you could squash all the commits into a single one (ideally also referencing #21300 in the message) and force push that, I'll approve and merge this.

The strangely out-of place multiplier that doesn't match its surroundings on `Set.subtracting.Seq.Empty.Box` reveals an older "bug". I've noticed that with same `1000` multiplier, the `Set.subtracting.Seq.Empty.Box` was faster than `Set.subtracting.Seq.Empty.Int` with #21300 applied (24 vs 133 μs on my ancient 💻), which was really odd. Turns out there's a minor inconsistency in the testing pattern and you've replicated it on the new `Seq` variant. As @lorentey explains in https://github.com/apple/swift/pull/18928#discussion_r212722518, the `identity` is used to prevent a future sufficiently smart compiler from hoisting the constant expression out of the loop. But he forgot to use that pattern on the `Box` variants of [`subtracting` and `intersection`](https://github.com/apple/swift/blob/11b290c4523d01bf9b93ce8e38c3127b5963314a/benchmark/single-source/SetTests.swift#L358-L370). Please fix that in your `Seq.Box` variants (we'll deal with fixing the original Set-to-Set separately).

Gumichocopengin8 commented on benchmark/single-source/SetTests.swift:
```diff
@@ -665,91 +664,50 @@ public let SetTests = [
 
     BenchmarkInfo(
-    name: "Set.filter.Int50.16k",
+    name: "Set.FilterMatchEven.16k",
```
Yeah, it does make sense.
I prefer `Set.filter.Int50.16k` because it follows other methods' names.
I thought you preferred `FilterMatchEven`.

Gumichocopengin8 commented:
@palimondo 
I also found some unused methods, so I deleted them all.
Could you please check the code.

palimondo commented:
This looks good to me. Could you please squash the commits and force-push a single one?

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -55,6 +55,43 @@ let setQ: Set<Int> = {
+let arrayYZ = Array(half ..< size + half) // 200 ..< 600
+let arrayY = Array(half ..< size)         // 200 ..< 400
+
+// Two arrays with 100% overlap, but different bucket counts (let's not make it
```
This comment is no longer necessary here.

palimondo commented:
> These two runtime outliers are interesting:
> 
> * Set.isStrictSubset.Seq.Empty.Int
> * Set.isStrictSuperset.Seq.Int.Empty
> 
> because they indicate we might benefit from special casing the empty set in the _NativeSet overrides. @lorentey?

Karoy, I apologize for ever doubting you 🙇‍♂️, I have not immediately realized the #21300 already handles these beautifully. It pays just the cheap cost of creating the `_Bitset(capacity: self.bucketCount)` and will bail out of the `Sequence` consuming loop ASAP. cc @LucianoPAlmeida

Gumichocopengin8 commented:
@palimondo 
I'm stuck on squashing the commits because of merge branch. I have to change pick to s so many times.
Is there any good ways to squash commits?

palimondo commented:
@Gumichocopengin8, I'm sorry, but I can only point you to [StackOverflow's answer](https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git). I'm lazy to learn all the command line incantations myself and I'm using the [Git Tower](https://www.git-tower.com/help/mac/commit-history/interactive-rebase).

palimondo commented:
Um… since this is just a single file we are changing here, maybe it would be enough to do manual backup of [SetTests.swift], revert to the parent of f45fe0c3a70b5aa4230fdae6eaf88926ec0a1519, copy the file in, commit that and force push? But I'm talking way out of my area of expertise here…  

If this remains a problem, we can leave it as is. It's just a matter of good PR hygiene.

Gumichocopengin8 commented:
@palimondo 
I found out the way.
I reset the commit and applied patch of this PR.

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo reviewed (changes requested):
@Gumichocopengin8 We've overlooked one more workload multiplier. Could you please fix it and force-push again?

palimondo commented:
@eeckstein, @lorentey How soon can we be landing the #21300? This additional performance coverage is tuned (in terms of workload scaling) for those improvements. Otherwise we'll have 4 benchmarks with really suboptimal runtimes:

A recent addition (21 days ago in #23690) `Set.isStrictSubset.Int100` shows up as regression, because we are fixing up the workload multiplier to account for the new [short circuiting optimization](https://github.com/apple/swift/pull/21300/files#diff-c8ef6275ed4edb2d48e444ed6c67dbdbR1186).

Similarly, these rely on the newly optimized handling and currently run at over 30k μs:
* `Set.isStrictSuperset.Seq.Int0`
* `Set.isStrictSuperset.Seq.Int.Empty`
* `Set.isStrictSubset.Seq.Empty.Int`

Are you OK with landing it this way now?

Gumichocopengin8 committed:
Variants for Set Sequence Methods

Gumichocopengin8 commented:
@palimondo 
Do we need to force push commit? I heard we can squash commits when we merge them.

palimondo commented:
@Gumichocopengin8 Hmm… I see there is an option to do that in the dropdown next to `Merge pull request` button 💡. I've never used that before. Theoretically that might have worked... but I don't know who would be credited as author then and [according to this description](https://help.github.com/en/articles/about-pull-request-merges#squash-and-merge-your-pull-request-commits) this feature must be allowed for the repository (and I don't know if/how that's configured for `apple/swift`).

Just to clarify, it is not necessary to squash commits into single one for a PR to be merged here. It is in fact beneficial to have many small and semantically single-themed commits if you are doing non-trivial multi-step changes such as refactoring, to ease the reviewer's job. But it is good practice to clean up the fits and starts (all the reverted dead ends etc.) and not pollute the history with unnecessary distractions. So thank you again for your patience with me, you did excellent work here!

I wanted to wait for a green-light from @lorentey and @eeckstein (I'm assuming many Apple devs are quite busy preparing for WWDC these days), but I guess I could merge this myself and ask for forgiveness if they'd disagree with my judgement. But I have noticed something we should change before accepting this. I'll do a re-review later today.

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -135,10 +211,133 @@ public let SetTests = [
+    setUpFunction: { blackHole([setOCD, arrayOAB]) }),
+  BenchmarkInfo(
+    name: "Set.isStrictSuperset.Seq.Int25",
+    runFunction: { n in run_SetIsStrictSupersetSeqInt(setAB, arrayB, true, 50 * n) },
```
```suggestion
    runFunction: { n in run_SetIsStrictSupersetSeqInt(setB, arrayBC, false, 50 * n) },
```
* Analogously for `Box25`.
* Use `setY` and `arrayYZ` on `Int50` below.
* Adjust `setUpFunction`s to match.

Gumichocopengin8 replied:
https://github.com/apple/swift/pull/24156#discussion_r277914847
I think keep it true is better. https://github.com/apple/swift/pull/24156#discussion_r277914678

palimondo replied:
Karoy was commenting on a Set.Set variant, back there:
>   runFunction: { n in run_SetIsStrictSupersetInt(setB, setAB, false, 50 * n) },

 which would have otherwise [short-circuited on the count conditions](https://github.com/apple/swift/pull/21300/files#diff-c8ef6275ed4edb2d48e444ed6c67dbdbR1186) because `setB.count < setAB.count`.

You have to flip the result here, because B is not a super set of AB. We are trying to exercise the [early exit on line 606](https://github.com/apple/swift/pull/21300/files#diff-5e43cf325f04c0738a77b7b8eb412e51R606).

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -135,10 +211,133 @@ public let SetTests = [
+    setUpFunction: { blackHole([setOCD, arrayOAB]) }),
+  BenchmarkInfo(
+    name: "Set.isSuperset.Seq.Int25",
+    runFunction: { n in run_SetIsSupersetSeqInt(setAB, arrayB, true, 50 * n) },
```
```suggestion
    runFunction: { n in run_SetIsSupersetSeqInt(setB, arrayBC, false, 50 * n) },
```
* Analogously for `Box25`.
* Use `setY` and `arrayYZ` on `Int50` below.
* Adjust `setUpFunction`s to match.

palimondo replied:
Trying to hit early exit at line 707 in Set. (This is basically unmodified by #21300.)

palimondo reviewed (changes requested):
I think we should stack the deck a bit more in favor of #21300 to showcase how its optimizations don't have to consume the entire `Sequence` because they provide early exits as soon as all the required conditions are met. I did one more review pass, comparing the original `Set.Set` benchmarks with the the newly added `Set.Sequence`. You have faithfully translated the arguments for `isSubset` and `isStrictSubset` 👍. But to better showcase the optimizations, we should be comparing against the *next-one-up* arrays (they have same size and same intersection, but their set difference is at the end of the sequence, instead of the beginning):
* `Int25` - use `setB` with **`arrayBC`** (instead of `arrayAB`); analogously for `Box25`
* `Int50` - use `setY` with **`arrayYZ`** (instead of `arrayXY`).

This is similar argument swapping like @lorentey [requested in his review before](https://github.com/apple/swift/pull/24156#discussion_r277914678). 
(_We will **not** be changing the arguments on `Set.Set` variants to match these, because it would cause unnecessary performance changes in our long-term tracking._)

Generally we should always have the same size of sequence (400) on the right side for all of these `Seq` variants. So for the `isSuperset.Seq` and `isStrictSuperset.Seq` we also need to swap the arguments and expected results around a bit, again, to allow for early exit from the optimizations.

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -135,10 +211,133 @@ public let SetTests = [
+    setUpFunction: { blackHole([setOAB, arrayOCD]) }),
+  BenchmarkInfo(
+    name: "Set.isStrictSubset.Seq.Int25",
+    runFunction: { n in run_SetIsStrictSubsetSeqInt(setB, arrayAB, true, 50 * n) },
```
Trying to hit [early exit at line 586 in NativeSet](https://github.com/apple/swift/pull/21300/files#diff-5e43cf325f04c0738a77b7b8eb412e51R586).

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -98,6 +133,47 @@ public let SetTests = [
+    setUpFunction: { blackHole([setOAB, arrayOCD]) }),
+  BenchmarkInfo(
+    name: "Set.isSubset.Seq.Int25",
+    runFunction: { n in run_SetIsSubsetSeqInt(setB, arrayAB, true, 50 * n) },
```
Trying to hit [early exit at line 568 in NativeSet](https://github.com/apple/swift/pull/21300/files#diff-5e43cf325f04c0738a77b7b8eb412e51R568).

palimondo commented:
@Gumichocopengin8 Will you find a time to do the requested modifications today? If not, I can make them for you, if you want...

Gumichocopengin8 committed:
minor fixes

Gumichocopengin8 commented:
@palimondo I fixed them.

palimondo commented:
I don’t see the fixes on `Box25` and `Int50` variant mentioned in the comments.

Gumichocopengin8 commented:
@palimondo oh sorry. Let me fix them tonight.

palimondo committed:
[benchmark] Set.Seq showcase optimized early exits

palimondo commented:
No problem. Here you go.

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please smoke test

palimondo committed:
[benchmark] Set.subtracting.Seq increased workload

The workload multipliers need to be increased in order to accommodate the pending improvements from `lorentey:set-on-fire`, so that these benchmark don't drop to under 20 μs runtimes when those optimizations kick in.

These multiplier will differ from those used on Set.subtracting.Set by a factor of 5.

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please smoke test os x platform

Gumichocopengin8 commented:
@palimondo Thanks!

palimondo commented:
Regarding the concern about the [4 long running benchmarks from above](https://github.com/apple/swift/pull/24156#issuecomment-490274157), I asked @lorentey via email if there was something else, beside the missing benchmarks which this PR provides, that is holding the merge of those optimizations. His response:
> #21300 is not ready to merge yet; it needs to be updated to ensure ABI compatibility (it currently lacks availability declarations and runtime checks). 
While the enhancements are indeed very nice, it is not super urgent to get this landed. We can cherry pick the isStrictSuperset changes though; the O(n) → O(1) speedup is a trivial change.

palimondo commented:
I will merge this and run the benchmarks on #21300 to validate we have all multipliers set up correctly to get 20—1000 μs runtimes. Then I'll temporarily disable the following benchmarks until `set-on-fire`'s ready to merge:

* `Set.isStrictSubset.Int100`
* `Set.isStrictSubset.Seq.Empty.Int`
* `Set.isStrictSuperset.Seq.Int.Empty`

@Gumichocopengin8 Thank you for your great work!
